mitiq/pec/types/types.py

```diff
@@ -0,0 +1,446 @@
+            )
+
+        self._basis_elements = set(basis_elements)
+
+    @property
+    def elements(self):
+        return self._basis_elements
+
+    def all_qubits(self) -> Set[cirq.Qid]:
+        """Returns the set of qubits that basis elements act on."""
+        qubits = set()
+        for noisy_op in self._basis_elements:
+            qubits.update(set(noisy_op.qubits))
+        return qubits
+
+    def add(self, *basis_elements) -> None:
+        """Add elements to the NoisyBasis.
+
+        Args:
+            basis_elements: Sequence of basis elements as `NoisyOperation`s to
+                add to the current basis elements.
+        """
+        for noisy_op in basis_elements:
+            if not isinstance(noisy_op, NoisyOperation):
+                raise ValueError(
+                    "All basis elements must be of type `NoisyOperation`."
+                )
+            self._basis_elements.add(noisy_op)
+
+    def extend_to(self, qubits: Sequence[List[cirq.Qid]]) -> None:
+        """Extends each basis element to act on the provided qubits.
+
+        Args:
+            qubits: Additional qubits for each basis element to act on.
+        """
+        for noisy_op in tuple(self._basis_elements):
+            self._basis_elements.update(
+                set(
+                    NoisyOperation.on_each(
+                        noisy_op.ideal_circuit(return_type="cirq"),
+                        noisy_op.real_matrix,
+                        qubits,
+                    )
+                )
+            )
+
+    def get_sequences(self, length: int) -> List[NoisyOperation]:
+        """Returns a list of all implementable NoisyOperation's of the given
+        length.
+
+        Example: If the ideal operations of the noisy basis elements are {I, X}
+            and length = 2, then this method returns the four NoisyOperations
+            whose ideal operations are {II, IX, XI, XX}.
+
+        Args:
+            length: Number of NoisyOperation's in each element of the returned
+                list.
+        """
+        sequences = []
+        for prod in product(self._basis_elements, repeat=length):
+            this_sequence = prod[0]
+            for noisy_op in prod[1:]:
+                this_sequence += noisy_op
+            sequences.append(this_sequence)
+        return sequences
+
+    def decompose(self, ideal: QPROGRAM):
+        raise NotImplementedError
+
+    def __len__(self):
+        return len(self._basis_elements)
+
+
+class OperationDecomposition:
+    def __init__(
+        self, ideal: QPROGRAM, basis_expansion: Dict[NoisyOperation, float]
+    ) -> None:
+        self._ideal = ideal
+
+        if not all(
+            isinstance(op, NoisyOperation) for op in basis_expansion.keys()
+        ):
+            raise TypeError(
+                "All keys of `basis_expansion` must be "
+                "of type `NoisyOperation`."
+            )
+
+        self._basis_expansion = cirq.LinearDict(basis_expansion)
```

Comment:
Is there any advantage in using a `cirq.LinearDict` instead of a standard Python `Dict`?

Reply:
A few - arithmetic (`ldict1 + ldict2`), string formatting

```python
ldict = cirq.LinearDict({cirq.X: 0.5, cirq.Z: -0.5})
print(ldict)
# 0.500*X-0.500*Z
```

etc. Time will tell if they're useful or not, but I suspect they will be.